app/routes.py

```diff
@@ -1,2 +1,257 @@
+    else:
+        tasks = Task.query.all()
+
+    tasks_response = []
+    for task in tasks:
+        tasks_response.append(task_dict(task))
+    return jsonify(tasks_response), 200
+        
+@tasks_bp.route("", methods=["POST"])
+def create_task():      
+    request_body = request.get_json()
+
+    if "title" not in request_body or "description" not in request_body \
+        or "completed_at" not in request_body:
+        return jsonify({"details" : "Invalid data"}), 400
+    
+    new_task = Task(
+        title=request_body["title"], 
+        description=request_body["description"], 
+        completed_at=request_body["completed_at"])
+
+    db.session.add(new_task)
+    db.session.commit()
+
+    response_body = {"task": task_dict(new_task)}
+    return jsonify(response_body), 201
+
+
+@tasks_bp.route("/<task_id>", methods=["GET"])
+def get_one_saved_task(task_id):
+    task = Task.query.get(task_id)
+    if task is None:
+        return jsonify(None), 404
+
+    if task.goal_id:
+        response_body = {"task": task_with_goal_dict(task)}
+        return jsonify(response_body), 200
+    return jsonify({"task": task_dict(task)}), 200
+
+@tasks_bp.route("/<task_id>", methods=["PUT"])
+def update_one_task(task_id):
+    task = Task.query.get(task_id)
+    if task is None:
+        return jsonify(None), 404
+
+    request_body = request.get_json()
+    task.title = request_body["title"]
+    task.description = request_body["description"]
+
+    db.session.commit()
+
+    response_body = {"task": task_dict(task)}
+    return jsonify(response_body), 200
+
+@tasks_bp.route("/<task_id>", methods=["DELETE"])
+def delete_one_task(task_id):
+    task = Task.query.get(task_id)
+    if task is None:
+        return jsonify(None), 404
+
+    db.session.delete(task)
+    db.session.commit()
+
+    return jsonify({
+        "details": f'Task {task.task_id} "{task.title}" successfully deleted'}), 200
+
+
+@tasks_bp.route("/<task_id>/mark_complete", methods=["PATCH"])
+def mark_one_task_complete(task_id):
+    task = Task.query.get(task_id)
+    if task is None:
+        return jsonify(None), 404
+
+    task.completed_at = datetime.now()
+
+    db.session.commit()
+
+    message = f"Someone just completed the task {task.title}."
+    slack_response(message)
+
+    response_body = {"task": task_dict(task)}
+    return jsonify(response_body), 200
+
+
+@tasks_bp.route("/<task_id>/mark_incomplete", methods=["PATCH"])
+def mark_one_task_incomplete(task_id):
+    task = Task.query.get(task_id)
+    if task is None:
+        return jsonify(None), 404
+
+    task.completed_at = None
+
+    db.session.commit()
+
+    response_body = {"task": task_dict(task)}
+    return jsonify(response_body), 200
+
+
+# using goals endpoint
+@goals_bp.route("", methods=["GET"])
+def get_saved_goals():
+    goals = Goal.query.all()
+
+    goals_response = []
+    for goal in goals:
+        goals_response.append(goal_dict(goal))
```

Comment:
We can list comprehension here

```suggestion
    goals_response = [ goal_dict(goal) for goal in goals ]
```